File: Scripts/DeviceSelectDefault.js
/* This script sets the value of Device Selection combobox widget. Useful
   when the main OPI is opened with DEVICE='valid_default_device' macro.
   
   Without this script the OPI with a specified DEVICE macro would still
   work, but the combobox widget would be empty and the user perhaps
   slightly confused. This is purely for cosmetics, but it's nice to have. */

importPackage(org.csstudio.display.builder.runtime.script);
importPackage(org.epics.vtype);

var device_macro = widget.getMacrosOrProperties().getValue("DEVICE");

var pv = PVUtil.getString(pvs[0]);

var pvstr =  widget.getPropertyValue("pv_value").getValue();
//java.lang.System.out.println("DEBUG: widget pv_value == " + pvstr);

if ((pvstr == "" || pvstr == "0") && device_macro != null) {
	/* This sets the value of the widget, but not the PV associated with it. */
	widget.setPropertyValue("pv_value", VString.of(device_macro,
                                                       Alarm.disconnected(), Time.now()));
}

File: Scripts/DeviceSelectDefault.py
from org.csstudio.display.builder.runtime.script import PVUtil
from org.csstudio.display.builder.runtime.script import ScriptUtil

logger = ScriptUtil.getLogger()
pv = PVUtil.getString(pvs[0])
if pv and pv != "":

    device_list = epik8sutil.conf_to_dev(widget)

    device_actual = widget.getEffectiveMacros().getValue("DEVICE")
    logger.info("SELECTION:"+pv)
    for dev in device_list:
        if pv == dev["NAME"]:
            logger.info("Found device: " + str(dev))
            widget.setPropertyValue("file","")
            widget.getPropertyValue("macros").add("DEVICE", dev["P"])
            widget.getPropertyValue("macros").add("HW", dev["TYPE"])
            widget.getPropertyValue("macros").add("CAM", dev["R"])
            widget.setPropertyValue("file","Camera_Main.bob")


